### ISSUE TITLE
negative int64 values handling is not compatible with reference Protobuf encoding

### ISSUE THREAD

**Issue by mourner** [mafintosh/protocol-buffers#36]
#14 by @yhahn had a nice int64 decoding/encoding fix, but the resulting byte sequence is not compatible with how it's handled in the reference Protobuf implementation. `int64` are cast into `uint64` using [Two's complement](http://en.wikipedia.org/wiki/Two%27s_complement), so when decoding, each byte should be flipped, and `1` added to the resulting number.

Here's how I hacked this together in my `pbf` implementation for decoding, and it seems to work: https://github.com/mapbox/pbf/blob/int64/index.js#L119-L120

**Comment by mafintosh** [comment]
@mourner your link 404s - i'd like to get this fixed there as well so could you update the link? :)

**Comment by mourner** [comment]
@mafintosh yeah, here's the method that works properly: https://github.com/mapbox/pbf/blob/master/index.js#L108-L125

**Comment by mafintosh** [comment]
@mourner is there any code for encoding 64bit varints as well?

**Comment by mourner** [comment]
@mafintosh no, I didn't do encoding yet, since int64's are pretty rare and usually a bad idea (vs varints), and we needed decoding only for compatibility with old pbfs. But I think it's pretty easy to write based on decoding.

**Comment by 0x7f** [comment]
Hi! I experienced the same issue and I think the same applies to `int32` as well. But only when encoding the value with google's c++ protobuf library and decoding with your library. The other way round seems to ok.

google c++ -> protocol-buffers:
```
$ diff -Naur <(./encoder | ./decoder) <(./encoder | node decoder.js)
--- /dev/fd/63  2015-07-15 09:41:32.874002144 +0000
+++ /dev/fd/62  2015-07-15 09:41:32.870002328 +0000
@@ -1,4 +1,4 @@
-as_int32 = -60
+as_int32 = 18446744069414584000
 as_sint32 = -60
-as_int64 = -60
+as_int64 = -68
 as_sint64 = -60
```

protocol-buffers -> google c++:
```
$ diff -Naur <(node encoder.js | ./decoder) <(node encoder.js | node decoder.js)
--- /dev/fd/63  2015-07-15 09:43:54.047495782 +0000
+++ /dev/fd/62  2015-07-15 09:43:54.047495782 +0000
@@ -1,4 +1,4 @@
 as_int32 = -60
 as_sint32 = -60
-as_int64 = -68
+as_int64 = -60
 as_sint64 = -60
```

I created a gist with all the code: https://gist.github.com/0x7f/2b94578a17246c2d0aa0
